From dfed9ef393338512ccad73e4e420437b6315c68b Mon Sep 17 00:00:00 2001
From: Steve deRosier <steve.derosier@lairdtech.com>
Date: Thu, 10 Nov 2016 16:50:48 -0800
Subject: [PATCH] laird: ath6kl: add module parameter to disable the fw debug
 log

The firmware host interest option register gives us an option to disable
the firmware from sending the debug logs. If we don't disable the debug
logs, we get constant information over the SDIO interface. By disabling the
logging, we reduce that information and can increase throughput in some
cases.

This patch adds a module parameter, disable_fw_dbglog, to do exactly this.
A user would set it to 1 and then when the firmware is loaded it will be
instructed to disable the firmware logs via the HI_OPTION_DISABLE_DBGLOG
being set in the host-interest flags.

The existing drivers don't send the flag down and I wanted to keep the
default behavior by having the module parameter specify the disable
instead of explicitly requiring the user to enable the debug. This
maintains consistent expected behavior unless a user expressly chooses to
change the behavior by setting the parameter.

In testing with Laird's WB50, an ar6004 based device, this results in a
basic throughput increase of 1 Mbps, or approximately 3.5% in my test
environment.

Bug: 10410
---
 drivers/net/wireless/ath/ath6kl/core.c   | 4 ++++
 drivers/net/wireless/ath/ath6kl/core.h   | 1 +
 drivers/net/wireless/ath/ath6kl/init.c   | 4 ++++
 drivers/net/wireless/ath/ath6kl/target.h | 1 +
 4 files changed, 10 insertions(+)

diff --git a/drivers/net/wireless/ath/ath6kl/core.c b/drivers/net/wireless/ath/ath6kl/core.c
index d3ccc23f1099..923b0f72b421 100644
--- a/drivers/net/wireless/ath/ath6kl/core.c
+++ b/drivers/net/wireless/ath/ath6kl/core.c
@@ -36,6 +36,7 @@ static unsigned int suspend_mode;
 static unsigned int wow_mode;
 static unsigned int uart_debug;
 static unsigned int uart_rate = 115200;
+static bool disable_fw_dbglog = false;
 static unsigned int ath6kl_p2p;
 static unsigned int testmode;
 static unsigned int recovery_enable;
@@ -48,12 +49,14 @@ module_param(suspend_mode, uint, 0644);
 module_param(wow_mode, uint, 0644);
 module_param(uart_debug, uint, 0644);
 module_param(uart_rate, uint, 0644);
+module_param(disable_fw_dbglog, bool, 0444);
 module_param(ath6kl_p2p, uint, 0644);
 module_param(testmode, uint, 0644);
 module_param(recovery_enable, uint, 0644);
 module_param(heart_beat_poll, uint, 0644);
 module_param(btcoex_chip_type, byte, 0444);
 module_param(btcoex_ant_config, byte, 0444);
+MODULE_PARM_DESC(disable_fw_dbglog, "Disable the firmware debug log");
 MODULE_PARM_DESC(recovery_enable, "Enable recovery from firmware error");
 MODULE_PARM_DESC(heart_beat_poll,
 		 "Enable fw error detection periodic polling in msecs - Also set recovery_enable for this to be effective");
@@ -200,6 +203,7 @@ int ath6kl_core_init(struct ath6kl *ar, enum ath6kl_htc_type htc_type)
 
 	set_bit(FIRST_BOOT, &ar->flag);
 
+	ar->disable_fw_dbglog = disable_fw_dbglog;
 	ath6kl_debug_init(ar);
 
 	ret = ath6kl_init_hw_start(ar);
diff --git a/drivers/net/wireless/ath/ath6kl/core.h b/drivers/net/wireless/ath/ath6kl/core.h
index 0441aca2b94f..45e3327ad41e 100644
--- a/drivers/net/wireless/ath/ath6kl/core.h
+++ b/drivers/net/wireless/ath/ath6kl/core.h
@@ -901,6 +901,7 @@ struct ath6kl {
 
 	u8 btcoex_chip_type;
 	u8 btcoex_ant_config;
+	bool disable_fw_dbglog;
 
 	struct {
 		enum wmi_phy_mode phy_mode;
diff --git a/drivers/net/wireless/ath/ath6kl/init.c b/drivers/net/wireless/ath/ath6kl/init.c
index 0195f7e458dd..78448d821620 100644
--- a/drivers/net/wireless/ath/ath6kl/init.c
+++ b/drivers/net/wireless/ath/ath6kl/init.c
@@ -628,6 +628,10 @@ int ath6kl_configure_target(struct ath6kl *ar)
 
 	param |= (0 << HI_OPTION_MAC_ADDR_METHOD_SHIFT);
 	param |= (0 << HI_OPTION_FW_BRIDGE_SHIFT);
+	if (ar->disable_fw_dbglog) {
+		param |= (1 << HI_OPTION_DISABLE_DBGLOG);
+		ath6kl_info("disabling fw dbglog\n");
+	}
 
 	if (ath6kl_bmi_write_hi32(ar, hi_option_flag, param) != 0) {
 		ath6kl_err("bmi_write_memory for setting fwmode failed\n");
diff --git a/drivers/net/wireless/ath/ath6kl/target.h b/drivers/net/wireless/ath/ath6kl/target.h
index d5eeeae7711b..40cff49a3ea2 100644
--- a/drivers/net/wireless/ath/ath6kl/target.h
+++ b/drivers/net/wireless/ath/ath6kl/target.h
@@ -315,6 +315,7 @@ struct host_interest {
 #define HI_OPTION_NUM_DEV_SHIFT   0x9
 
 #define HI_OPTION_FW_BRIDGE_SHIFT 0x04
+#define HI_OPTION_DISABLE_DBGLOG  0X06
 
 /* Fw Mode/SubMode Mask
 |------------------------------------------------------------------------------|
-- 
2.11.0

